File: fixtures/dotnet/j10/Fixture.cs
// N+1 query example using Entity Framework Core pattern
// This causes 1 + N (items) + N (customer) queries for N orders

using System;
using System.Collections.Generic;
using System.Linq;
using Microsoft.EntityFrameworkCore;

// Simulated ORM with query counter
static class QueryCounter
{
    private static int _queryCount = 0;

    public static void Reset() { _queryCount = 0; }
    public static int GetCount() { return _queryCount; }
    public static void Increment() { _queryCount++; }
}

class Customer
{
    public int Id { get; set; }
    public string Name { get; set; }
    public string Email { get; set; }
}

class OrderItem
{
    public int Id { get; set; }
    public string Name { get; set; }
    public int Qty { get; set; }
    public double Price { get; set; }
}

class Order
{
    public int Id { get; set; }
    public int CustomerId { get; set; }
    public string Status { get; set; }

    // Navigation properties (lazy loaded by default)
    public virtual Customer Customer { get; set; }
    public virtual List<OrderItem> Items { get; set; }
}

// Simulated DbContext with N+1 problem
class AppDbContext : DbContext
{
    public DbSet<Order> Orders { get; set; }
    public DbSet<OrderItem> OrderItems { get; set; }
    public DbSet<Customer> Customers { get; set; }
}

class OrderRepository
{
    private readonly AppDbContext _db;

    public OrderRepository(AppDbContext db) { _db = db; }

    public List<Order> FindAllOrders()
    {
        QueryCounter.Increment();
        return new List<Order>
        {
            new Order { Id = 1, CustomerId = 1, Status = "shipped" },
            new Order { Id = 2, CustomerId = 2, Status = "pending" },
            new Order { Id = 3, CustomerId = 1, Status = "delivered" }
        };
    }

    public List<OrderItem> FindItemsByOrderId(int orderId)
    {
        QueryCounter.Increment(); // N+1: one query per order
        var itemsMap = new Dictionary<int, List<OrderItem>>
        {
            [1] = new List<OrderItem> { new OrderItem { Id = 1, Name = "Widget", Qty = 2, Price = 10 } },
            [2] = new List<OrderItem>
            {
                new OrderItem { Id = 2, Name = "Gadget", Qty = 1, Price = 25 },
                new OrderItem { Id = 3, Name = "Doohickey", Qty = 3, Price = 5 }
            },
            [3] = new List<OrderItem> { new OrderItem { Id = 4, Name = "Widget", Qty = 1, Price = 10 } }
        };
        return itemsMap.ContainsKey(orderId) ? itemsMap[orderId] : new List<OrderItem>();
    }

    public Customer FindCustomerById(int customerId)
    {
        QueryCounter.Increment(); // N+1: one query per order for customer
        var customers = new Dictionary<int, Customer>
        {
            [1] = new Customer { Id = 1, Name = "Alice", Email = "alice@example.com" },
            [2] = new Customer { Id = 2, Name = "Bob", Email = "bob@example.com" }
        };
        return customers.ContainsKey(customerId) ? customers[customerId] : null;
    }

    // Optimized: uses .Include() to eager load everything in a single query
    public List<Order> FindAllOrdersWithItemsAndCustomers()
    {
        QueryCounter.Increment(); // Optimized: single join query
        // In real EF Core: _db.Orders.Include(o => o.Items).Include(o => o.Customer).ToList()
        // Returns fully hydrated entities in one round-trip
        return new List<Order>();
    }
}

// N+1 problem: fetches orders, then for each order fetches items and customer separately
class OrderService
{
    private readonly OrderRepository _repo;

    public OrderService(OrderRepository repo) { _repo = repo; }

    public List<Dictionary<string, object>> GetOrdersWithDetails()
    {
        var orders = _repo.FindAllOrders();

        return orders.Select(order =>
        {
            var items = _repo.FindItemsByOrderId(order.Id);          // N queries
            var customer = _repo.FindCustomerById(order.CustomerId); // N queries
            return new Dictionary<string, object>
            {
                ["order"] = order,
                ["items"] = items,
                ["customer"] = customer
            };
        }).ToList();
    }
}

File: fixtures/java/j10/Fixture.java
// N+1 query example using JPA/Hibernate pattern
// This causes 1 + N (items) + N (customer) queries for N orders

import javax.persistence.*;
import java.util.*;
import java.util.stream.Collectors;

// Simulated ORM with query counter
class QueryCounter {
    private static int queryCount = 0;

    static void reset() { queryCount = 0; }
    static int getCount() { return queryCount; }
    static void increment() { queryCount++; }
}

@Entity
class Customer {
    @Id private int id;
    private String name;
    private String email;

    Customer() {}
    Customer(int id, String name, String email) {
        this.id = id; this.name = name; this.email = email;
    }

    int getId() { return id; }
    String getName() { return name; }
    String getEmail() { return email; }
}

@Entity
class OrderItem {
    @Id private int id;
    private String name;
    private int qty;
    private double price;

    OrderItem() {}
    OrderItem(int id, String name, int qty, double price) {
        this.id = id; this.name = name; this.qty = qty; this.price = price;
    }
}

@Entity
class Order {
    @Id private int id;
    private int customerId;
    private String status;

    @ManyToOne(fetch = FetchType.LAZY)
    private Customer customer;

    @OneToMany(fetch = FetchType.LAZY)
    private List<OrderItem> items;

    Order() {}
    Order(int id, int customerId, String status) {
        this.id = id; this.customerId = customerId; this.status = status;
    }

    int getId() { return id; }
    int getCustomerId() { return customerId; }
    String getStatus() { return status; }
    Customer getCustomer() { return customer; }
    List<OrderItem> getItems() { return items; }
}

// Simulated repository with N+1 problem
class OrderRepository {

    List<Order> findAllOrders() {
        QueryCounter.increment();
        return Arrays.asList(
            new Order(1, 1, "shipped"),
            new Order(2, 2, "pending"),
            new Order(3, 1, "delivered")
        );
    }

    List<OrderItem> findItemsByOrderId(int orderId) {
        QueryCounter.increment(); // N+1: one query per order
        Map<Integer, List<OrderItem>> itemsMap = new HashMap<>();
        itemsMap.put(1, Arrays.asList(new OrderItem(1, "Widget", 2, 10)));
        itemsMap.put(2, Arrays.asList(new OrderItem(2, "Gadget", 1, 25), new OrderItem(3, "Doohickey", 3, 5)));
        itemsMap.put(3, Arrays.asList(new OrderItem(4, "Widget", 1, 10)));
        return itemsMap.getOrDefault(orderId, Collections.emptyList());
    }

    Customer findCustomerById(int customerId) {
        QueryCounter.increment(); // N+1: one query per order for customer
        Map<Integer, Customer> customers = new HashMap<>();
        customers.put(1, new Customer(1, "Alice", "alice@example.com"));
        customers.put(2, new Customer(2, "Bob", "bob@example.com"));
        return customers.get(customerId);
    }

    // Optimized: uses JOIN FETCH to load everything in a single query
    List<Order> findAllOrdersWithItemsAndCustomers() {
        QueryCounter.increment(); // Optimized: single join query
        // In real JPA: @Query("SELECT o FROM Order o JOIN FETCH o.items JOIN FETCH o.customer")
        // Returns fully hydrated entities in one round-trip
        return Arrays.asList(
            // orders with items and customers pre-loaded
        );
    }
}

// N+1 problem: fetches orders, then for each order fetches items and customer separately
class OrderService {

    private final OrderRepository repo = new OrderRepository();

    List<Map<String, Object>> getOrdersWithDetails() {
        List<Order> orders = repo.findAllOrders();

        return orders.stream().map(order -> {
            List<OrderItem> items = repo.findItemsByOrderId(order.getId());       // N queries
            Customer customer = repo.findCustomerById(order.getCustomerId());     // N queries
            Map<String, Object> enriched = new HashMap<>();
            enriched.put("order", order);
            enriched.put("items", items);
            enriched.put("customer", customer);
            return enriched;
        }).collect(Collectors.toList());
    }
}
